pvlib/iotools/solcast.py

```diff
@@ -0,0 +1,374 @@
+""" Functions to access data from the Solcast API.
+"""
+
+import requests
+import pandas as pd
+from dataclasses import dataclass
+
+
+BASE_URL = "https://api.solcast.com.au/data"
+
+@dataclass
+class ParameterMap:
+    solcast_name: str
+    pvlib_name: str
+    conversion: callable=lambda x: x
+
+# define the conventions between Solcast and PVLib nomenclature and units
+VARIABLE_MAP = [
+    ParameterMap("air_temp", "temp_air"),  # air_temp -> temp_air (deg C)
+    ParameterMap("surface_pressure", "pressure", lambda x: x*100),  # surface_pressure (hPa) -> pressure (Pa)
+    ParameterMap("dewpoint_temp", "temp_dew"),  # dewpoint_temp -> temp_dew (deg C)
+    ParameterMap("gti", "poa_global"),  # gti (W/m^2) -> poa_global (W/m^2)
+    ParameterMap("wind_speed_10m", "wind_speed"),  # wind_speed_10m (m/s) -> wind_speed (m/s)
+    ParameterMap("wind_direction_10m", "wind_direction"),  # wind_direction_10m (deg) -> wind_direction  (deg) (Convention?)
+    ParameterMap(
+        "azimuth", "solar_azimuth", lambda x: abs(x) if x <= 0 else 360 - x
+                 ),  # azimuth -> solar_azimuth (degrees) (different convention)
+    ParameterMap("precipitable_water", "precipitable_water", lambda x: x*10),  # precipitable_water (kg/m2) -> precipitable_water (cm)
+    ParameterMap("zenith", "solar_zenith")  # zenith -> solar_zenith
+]
+
+
+def get_solcast_tmy(
+    latitude, longitude, api_key, map_variables=True, **kwargs
+):
+    """Get the irradiance and weather for a Typical Meteorological Year (TMY) at a requested location.
+
+    Derived from satellite (clouds and irradiance over non-polar continental areas) and
+    numerical weather models (other data). The TMY is calculated with data from 2007 to 2023.
+
+    Parameters
+    ----------
+    latitude : float
+        in decimal degrees, between -90 and 90, north is positive
+    longitude : float
+        in decimal degrees, between -180 and 180, east is positive
+    api_key : str
+        To access Solcast data you will need an API key: https://toolkit.solcast.com.au/register.
+    map_variables: bool, default: True
+        When true, renames columns of the DataFrame to pvlib variable names
+        where applicable. See variable :const:`VARIABLE_MAP`.
+        Time is made the index with the "period mid" convention from Solcast's "period end".
+    kwargs:
+        Optional parameters passed to the API. See https://docs.solcast.com.au/ for full list of parameters.
+
+    Returns
+    -------
+    df : pandas.DataFrame
+        containing the values for the parameters requested

```

Comment:
```suggestion
    data : pandas.DataFrame
        containing the values for the parameters requested

```

Should also output a meta dictionary, even if empty